icu4c/source/common/locid.cpp

```diff
@@ -504,34 +508,1071 @@ Locale::operator==( const   Locale& other) const
+    void readTerritoryAlias(UResourceBundle* alias,
+                            UniqueCharStrings* strings,
+                            LocalMemory<const char*>& types,
+                            LocalMemory<int32_t>& replacementIndexes,
+                            int32_t &length, UErrorCode &status);
+
+    // Read the variantAlias data from alias to
+    // strings+types+replacementIndexes
+    // Allocate length itmes for types, to store the type field.
+    // Allocate length itmes for replacementIndexes,
+    // to store the index in the strings for the replacement variant.
+    void readVariantAlias(UResourceBundle* alias,
+                          UniqueCharStrings* strings,
+                          LocalMemory<const char*>& types,
+                          LocalMemory<int32_t>& replacementIndexes,
+                          int32_t &length, UErrorCode &status);
+};
+
+/**
+ * A class to hold the Alias Data.
+ */
+class AliasData {
+public:
+    static const AliasData* singleton(UErrorCode status) {
+        umtx_initOnce(AliasData::gInitOnce, &AliasData::loadData, status);
+        return gSingleton;
+    }
+
+    const CharStringMap* languageMap() const {
+      return language.getAlias();
+    }
+
+    const CharStringMap* scriptMap() const {
+      return script.getAlias();
+    }
+
+    const CharStringMap* territoryMap() const {
+      return territory.getAlias();
+    }
+
+    const CharStringMap* variantMap() const {
+      return variant.getAlias();
+    }
+
+    static void U_CALLCONV loadData(UErrorCode &status);
+    static UBool U_CALLCONV cleanup();
+
+    static UInitOnce gInitOnce;
+
+private:
+    AliasData(CharStringMap* languageMap,
+              CharStringMap* scriptMap,
+              CharStringMap* territoryMap,
+              CharStringMap* variantMap,
+              UniqueCharStrings* strings)
+        : language(languageMap),
+          script(scriptMap),
+          territory(territoryMap),
+          variant(variantMap),
+          strings(strings) {
+    }
+
+    static const AliasData* gSingleton;
+
+    LocalPointer<const CharStringMap> language;
+    LocalPointer<const CharStringMap> script;
+    LocalPointer<const CharStringMap> territory;
+    LocalPointer<const CharStringMap> variant;
+    LocalPointer<const UniqueCharStrings> strings;
+
+    friend class AliasDataBuilder;
+};
+
+
+const AliasData* AliasData::gSingleton = nullptr;
+UInitOnce AliasData::gInitOnce = U_INITONCE_INITIALIZER;
+
+UBool U_CALLCONV
+AliasData::cleanup()
+{
+    gInitOnce.reset();
+    delete gSingleton;
+    return TRUE;
+}
+
+/**
+ * Read the languageAlias data from alias to strings+types+replacementIndexes.
+ * Allocate length itmes for types, to store the type field. Allocate length
+ * itmes for replacementIndexes, to store the index in the strings for the
+ * replacement language.
+ */
+void
+AliasDataBuilder::readLanguageAlias(
+        UResourceBundle* alias,
```

Comment:
This works, but I find it much cleaner to work with ures_getValueWithFallback() and ResourceDataValue etc.

Is there a reason you didn't copy that kind of data-loading code from loclikelysubtags.cpp? It should have been possible to copy more of that code, incl. readStrings() etc.

Using the public API will work, and there is no time to change this now.

Maybe a TODO?

Reply:
Because I copy code from somewhere else (Region::loadRegionData i18n/region.cpp. That file read from the same resource file so I copy the code from there instead from loclikelysubtags.cpp which read a different resource file.) No way I can tell which one is better than the other the moment I copied the code, but it is clear copy from i18n/region.cpp is easier since it read the same resource file.